tests/test_user_context.py

```diff
@@ -1296,3 +1298,590 @@ def test_decide_experiment(self):
+
+        expected_reasons = [
+            'Evaluating audiences for experiment "group_exp_2": [].',
+            'Audiences for experiment "group_exp_2" collectively evaluated to TRUE.',
+            'User "test_user" is in experiment group_exp_2 of group 19228.',
+            'User "test_user" is in variation "group_exp_2_control" of experiment group_exp_2.'
+        ]
+
+        self.assertEqual(decide_decision.reasons, expected_reasons)
+
+    def test_invalid_delivery_rule_return_decision__forced_decision(self):
+        """
+        Should return valid decision after setting invalid delivery rule variation in forced decision.
+        """
+        opt_obj = optimizely.Optimizely(json.dumps(self.config_dict_with_features))
+        user_context = OptimizelyUserContext(opt_obj, "test_user", {})
+
+        context = OptimizelyUserContext.OptimizelyDecisionContext('test_feature_in_rollout', '211127')
+        decision = OptimizelyUserContext.OptimizelyForcedDecision('invalid')
+
+        status = user_context.set_forced_decision(context, decision)
+        self.assertTrue(status)
+        status = user_context.get_forced_decision(context)
+        self.assertEqual(status.variation_key, 'invalid')
+
+        decide_decision = user_context.decide('test_feature_in_rollout', ['INCLUDE_REASONS'])
+        self.assertEqual(decide_decision.variation_key, '211149')
+        self.assertEqual(decide_decision.rule_key, '211147')
+        self.assertTrue(decide_decision.enabled)
+        self.assertEqual(decide_decision.flag_key, 'test_feature_in_rollout')
+        self.assertEqual(decide_decision.user_context.user_id, 'test_user')
+        self.assertEqual(decide_decision.user_context.get_user_attributes(), {})
+        self.assertTrue(set(decide_decision.reasons).issuperset(set([
+            'Invalid variation is mapped to flag (test_feature_in_rollout), '
+            'rule (211127) and user (test_user) in the forced decision map.'
+        ])))
+
+    def test_invalid_experiment_rule_return_decision__forced_decision(self):
+        """
+        Should return valid decision after setting invalid experiment
+        rule variation in forced decision.
+        """
+        opt_obj = optimizely.Optimizely(json.dumps(self.config_dict_with_features))
+        user_context = OptimizelyUserContext(opt_obj, "test_user", {})
+
+        context = OptimizelyUserContext.OptimizelyDecisionContext('test_feature_in_experiment_and_rollout',
+                                                                  'group_exp_2')
+        decision = OptimizelyUserContext.OptimizelyForcedDecision('invalid')
+
+        status = user_context.set_forced_decision(context, decision)
+        self.assertTrue(status)
+        status = user_context.get_forced_decision(context)
+        self.assertEqual(status.variation_key, 'invalid')
+
+        decide_decision = user_context.decide('test_feature_in_rollout', ['INCLUDE_REASONS'])
```

Comment:
@ozayr-zaviar Why is here a different flag on decision than in the context?
Is this intentional? Can you explain?